service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/UpdateControlledBigQueryDatasetsLifetimeStep.java

```diff
@@ -0,0 +1,102 @@
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    var workingMap = context.getWorkingMap();
+    validateRequiredEntries(
+        workingMap,
+        CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP,
+        CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP,
+        CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP);
+
+    Map<UUID, String> resourceIdToDefaultTableLifetimeMap =
+        workingMap.get(
+            CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP,
+            new TypeReference<>() {});
+    Map<UUID, String> resourceIdToDefaultPartitionLifetimeMap =
+        workingMap.get(
+            CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP,
+            new TypeReference<>() {});
+
+    Map<UUID, String> resourceIdsToWorkspaceIdMap =
+        workingMap.get(CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP, new TypeReference<>() {});
+    List<ControlledBigQueryDatasetResource> updatedResources = new ArrayList<>();
+
+    List<ControlledResource> controlledBigQueryDatasets =
+        context
+            .getWorkingMap()
+            .get(CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {});
+
+    assert controlledBigQueryDatasets != null;
+    assert resourceIdToDefaultPartitionLifetimeMap != null;
+    assert resourceIdToDefaultTableLifetimeMap != null;
+    assert resourceIdsToWorkspaceIdMap != null;
+
+    for (var resource : controlledBigQueryDatasets) {
+      var id = resource.getResourceId();
+      if (resourceIdsToWorkspaceIdMap.containsKey(id)) {
+        boolean updated =
+            resourceDao.updateBigQueryDatasetDefaultTableAndPartitionLifetime(
+                resource.castByEnum(CONTROLLED_GCP_BIG_QUERY_DATASET),
+                Long.valueOf(resourceIdToDefaultTableLifetimeMap.get(id)),
+                Long.valueOf(resourceIdToDefaultPartitionLifetimeMap.get(id)));
+
+        if (updated) {
+          updatedResources.add(
+              resourceDao
+                  .getResource(UUID.fromString(resourceIdsToWorkspaceIdMap.get(id)), id)
+                  .castToControlledResource()
+                  .castByEnum(CONTROLLED_GCP_BIG_QUERY_DATASET));
+        }
+      }
+    }
+
+    setResponse(context, updatedResources, HttpStatus.OK);
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    List<ControlledBigQueryDatasetResource> controlledBigQueryDatasets =
+        context
+            .getWorkingMap()
+            .get(CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {});
+
+    if (controlledBigQueryDatasets == null || controlledBigQueryDatasets.isEmpty()) {
+      return StepResult.getStepResultSuccess();
+    }
+    for (var resource : controlledBigQueryDatasets) {
+      resourceDao.updateBigQueryDatasetDefaultTableAndPartitionLifetime(resource, null, null);
```

Comment:
Oh, good catch! So, since we iterate through the BQ datasets from the initial `listControlledBigQueryDatasetsWithoutLifetime` query, we can use those lifetime values for our undo.